File: leetcode_solutions/0042-trapping-rain-water/solution.java
class Solution {
    public int trap(int[] height) {
        int rPtr = 0;
        int lPtr = height.length - 1;

        int rMax = height[rPtr];
        int lMax = height[lPtr];

        int totalWater = 0;
        while (rPtr < lPtr){
            if (height[rPtr] < height[lPtr]){
                if (height[rPtr] >= rMax){
                    rMax = height[rPtr];
                } else {
                    totalWater += rMax - height[rPtr];
                }
                rPtr++;
            } else {
                if (height[lPtr] >= lMax){
                    lMax = height[lPtr];
                } else {
                    totalWater += lMax - height[lPtr];
                }
                lPtr--;
            }
        }
        return totalWater;
    }
}

File: leetcode_solutions/0042-trapping-rain-water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left_max = height[0]
        right_max = height[len(height) - 1]
        left_ptr = 0
        right_ptr = len(height) - 1
        water = 0

        while (left_ptr < right_ptr):
            if (height[left_ptr] <= height[right_ptr]):
                if (height[left_ptr] < left_max):
                    water += left_max - height[left_ptr]
                else:
                    left_max = height[left_ptr]
                left_ptr += 1
            else:
                if (height[right_ptr] < right_max):
                    water += right_max - height[right_ptr]
                else:
                    right_max = height[right_ptr]
           
                right_ptr -= 1
        return water
